Modules/SubModule_Import/xASL_imp_DCM2NII_Subject_ShuffleTheDynamics.m

```diff
@@ -130,40 +132,59 @@
         end
     end
     
-    %% FME (Hadamard) Check
-    if numel(nii_files)>=1
+    %% Hadamard Check
+    
+    bHadamard = false;
+    bHadamardFME = false;
+    
+    % Determine if we have a Hadamard sequence based on the parameters of the studyPar.json
+    if xASL_exist(x.dir.studyPar,'file')
+        studyPar = spm_jsonread(x.dir.studyPar);
+        if isfield(studyPar,'HadamardMatrixType')
+            bHadamard = true;
+        end
+    end
+    
+    
+    %% Check for specific Hadamard sequences and reorder the NIfTI volumes if necessary
+    if numel(nii_files)>=1 
         [resultPath, resultFile] = xASL_fileparts(nii_files{1});
         % Check if we have the corresponding JSON file
-        if exist(fullfile(resultPath, [resultFile '.json']), 'file')
+        if exist(fullfile(resultPath, [resultFile '.json']), 'file') 
         	% Load the JSON
             resultJSON = spm_jsonread(fullfile(resultPath, [resultFile '.json']));
             % Check if we have the SeriesDescription field
-            if isfield(resultJSON,'SeriesDescription')
-            	% Determine if we have a Hadamard encoded ASL scan
-                isHadamardFME = ~isempty(regexp(char(resultJSON.SeriesDescription),'(Encoded_Images_Had)\d\d(_)\d\d(_TIs_)\d\d(_TEs)', 'once'));
-                if isHadamardFME
-                    % Check image
-                    if exist(nii_files{1} ,'file')
-                    	% Determine the number of time points within each NIfTI
-                        imASL = xASL_io_Nifti2Im(nii_files{1});
-						numberTEs = length(resultJSON.EchoTime);
-                        numberPLDs = int32(size(imASL,4)/numberTEs);
-						
-						% Reorder TEs and PLDs - first cycle TE afterwards PLD
-						vectorOldOrder = zeros(size(imASL,4),1);
-						for iPLD = 1:(double(numberPLDs))
-							vectorOldOrder((1:numberTEs)+(iPLD-1)*numberTEs) = (iPLD-1)+1:numberPLDs:size(imASL,4);
-						end
-						imASL(:,:,:,1:end) = imASL(:,:,:,vectorOldOrder);
-						xASL_io_SaveNifti(nii_files{1},nii_files{1},imASL);
-						% Repeat Echo Times
-						resultJSON.EchoTime = repmat(resultJSON.EchoTime,numberPLDs,1);
-						% Save the JSON with the updated echo times
-						spm_jsonwrite(fullfile(resultPath, [resultFile '.json']),resultJSON);
-                    else
-                    	% Fallback (something went wrong)
-						warning('FME Hadamard sequence with 1 PLD only');
+            if isfield(resultJSON,'SeriesDescription') || bHadamard
```

Comment:
Only here the `bHadamard` is used, so the reader has to read 8 lines before it is clear that this part is for Hadamard (an exception!) only. Make this clear at the start of this section, by adding a section title (comment) and moving the `bHadamard` check to the beginning of this code section.

Reply:
Doesn't work. Check my PR comment above and the suggestion on how to improve it.